src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -1799,6 +1800,29 @@ private fun checkEmptyFunctionReturnType(holder: RsAnnotationHolder, fn: RsFunct
     }
 }
 
+private fun checkRecursiveAsyncFunction(holder: RsAnnotationHolder, fn: RsFunction) {
+    if (!fn.isAsync) return
+    val recursiveCall = fn
+        .descendantsOfType<RsFieldLookup>()
+        .firstOrNull {
+            if (!it.isAsync) return@firstOrNull false
+            val dotExpr = it.parent as? RsDotExpr
+            val callExpr = dotExpr?.expr as? RsCallExpr
+            val pathExpr = callExpr?.expr as? RsPathExpr ?: return@firstOrNull false
+            val path = pathExpr.path
+            !path.hasColonColon && path.reference?.resolve() == fn
+        } ?: return
+    if (fn.hasAsyncRecursionProcMacro()) return
+    RsDiagnostic.RecursiveAsyncFunction(recursiveCall).addToHolder(holder)
+}
+
+private fun RsFunction.hasAsyncRecursionProcMacro(): Boolean {
+    val attr = ProcMacroAttribute
+        .getProcMacroAttributeWithoutResolve(this, ignoreProcMacrosDisabled = true)
```

Comment:
The javadoc for `getProcMacroAttributeWithoutResolve` methods says
```
Differs from [getProcMacroAttribute] in that it does not take into account hardcoded macro attributes
```
and `async_recursion` macro is included in the hardcoded macro attributes list.

Does this mean that we should use `getProcMacroAttribute` here instead? Maybe I'm missing something, or the javadoc could describe the difference in a bit more detail.
Cc @vlad20012

Reply:
We need to use `getProcMacroAttributeWithoutResolve` here, because `getProcMacroAttribute` returns `None` for `async_recursion` (we threat it as identity macro)

Reply:
Perhaps it's better to just remove `async_recursion` from the hardcoded macro list